modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/queue/ViewableQueue.scala

```diff
@@ -0,0 +1,51 @@
+package org.tessellation.node.shared.domain.queue
+
+import cats.effect.std.{Mutex, Queue}
+import cats.effect.{Concurrent, Ref}
+import cats.implicits._
+
+trait ViewableQueue[F[_], A] {
+  def offer(item: A): F[Unit]
+
+  def take: F[A]
+
+  def tryOfferN(items: List[A]): F[List[A]]
+
+  def tryTakeN(n: Option[Int]): F[List[A]]
+
+  def view: F[List[A]]
+}
+
+object ViewableQueue {
+
+  def unbounded[F[_]: Concurrent, A]: F[ViewableQueue[F, A]] =
+    Queue.unbounded[F, A].flatMap(q => make(q))
+
+  def make[F[_]: Concurrent, A](queue: Queue[F, A]): F[ViewableQueue[F, A]] =
+    (
+      Ref[F].of(List.empty[A]),
+      Mutex[F]
+    ).tupled.map {
+      case (state, mutex) =>
+        new ViewableQueue[F, A] {
```

Comment:
We should avoid locking if possible and write lock-free code or without semantically blocking.
`tryOfferN` and `tryTakeN` implementations on Queue take this approach, hence their name `try...`. By wrapping it basically by mutex, it's no longer `try` and it's blocking now. Just thinking out loud (didn't validate if it works) maybe you can keep the queue inside the Ref, so you use the Ref's internal mechanism for retrying in case of concurrent update. Or maybe we should not use a `Queue` anymore, if we need to see all the data anyway.

Reply:
@marcinwadon I attempted to write a test to explore this concern by comparing a blocking and non-blocking version of `ViewableQueue`. Indeed, I am unable to get the `Ref` and `Queue` out of sync (although they do not share the same order), so it would seem that the non-blocking version would be preferred. However, I was curious to explore timing differences between the two implementations and found that the blocking implementation may be more performant (at least in this particular scenario). That was unexpected so perhaps the test is not well suited but in any case here are the two tests with results. 

- With 20k Int, non-blocking, I get an elapsed time of "Test took 8843498 microseconds."
- With 20k Int, blocking, I get an elapsed time of "Test took 2255935 microseconds."

Note that I ran the tests independently to get a more accurate estimate of timing. If you run them at once their timing will be nearly identical.

## Test
```Scala
object ViewableQueueSuite extends SimpleIOSuite with Checkers {

//  test("Insert 20k Int to blocking Queue") {
//    for {
//      start <- Clock[IO].realTime
//      queue <- Queue.unbounded[IO, Int]
//      viewQueue <- ViewableQueue.makeBlocking[IO, Int](queue)
//      _ <- List.range(0, 20000).map(i => viewQueue.offer(i)).parSequence
//      result <- (viewQueue.view, viewQueue.tryTakeN(Some(20000))).parTupled
//      (viewResult, takeResult) = result
//      end <- Clock[IO].realTime
//      _ <- IO.println(s"Test took ${end - start}.")
//    } yield expect(viewResult.size == takeResult.size)
//  }

  test("Insert 20k Int to NON-blocking Queue") {
    for {
      start <- Clock[IO].realTime
      queue <- Queue.unbounded[IO, Int]
      viewQueue <- ViewableQueue.makeNonblocking[IO, Int](queue)
      _ <- List.range(0, 20000).map(i => viewQueue.offer(i)).parSequence
      result <- (viewQueue.view, viewQueue.tryTakeN(Some(20000))).parTupled
      (viewResult, takeResult) = result
      end <- Clock[IO].realTime
      _ <- IO.println(s"Test took ${end - start}.")
    } yield expect(viewResult.size == takeResult.size)
  }
}
```

Reply:
I think you should measure the time of the particular method only, not the initialization and numbers generation. Anyway, the test doesn't check the concurrent access which would happen in real life

Reply:
Do you have any suggestions on ways I can stress test this? I have made an attempt with the test below and find that the blocking version is approx 4x faster than the non-blocking version for 20k entries inserted via 100 Fibers each inserting 200 elements with small randomized delay between each element.

**Result**
Blocking -------- Total insertion time for all fibers: 2603593 microseconds
Non-blocking - Total insertion time for all fibers: 7764411 microseconds

**Test**
```Scala
  test("Insert 20k Int to NON-blocking Queue via multiple fibers") {
    for {
      queue <- Queue.unbounded[IO, Int]
      viewQueue <- ViewableQueue.makeNonblocking[IO, Int](queue)
      actions = List
        .range(0, 100)
        .map(i => List.range(i * 200, (i + 1) * 200))
        .flatMap(list => list.map(item => IO.sleep((math.random() * 2).millisecond) *> viewQueue.offer(item)))
      start <- Clock[IO].realTime
      fibers <- actions.parTraverseN(10)(_.start)
      _ <- fibers.traverse(_.join)
      end <- Clock[IO].realTime
      _ <- IO.println(s"Total insertion time for all fibers: ${end - start}")
      result <- (viewQueue.view, viewQueue.tryTakeN(Some(20000))).tupled
      (viewResult, takeResult) = result
    } yield expect(viewResult.size == takeResult.size)
  }
````

Reply:
@scasplte2 Can you try this instead?

```scala
object ViewableQueue {

  def make[F[_]: Concurrent, A]: F[ViewableQueue[F, A]] =
    Ref[F].of(List.empty[A]).map { state =>
      new ViewableQueue[F, A] {
        def offer(item: A): F[Unit] =
          state.update(_ :+ item)

        def take: F[A] =
          state.modify {
            case head :: tail => (tail, head)
            case Nil => throw new NoSuchElementException("take from an empty queue")
          }

        def tryOfferN(items: List[A]): F[List[A]] =
          state.modify { current =>
            val updated = current ++ items
            (updated, List.empty[A])
          }

        def tryTakeN(n: Option[Int]): F[List[A]] =
          state.modify { current =>
            val (taken, remaining) = current.splitAt(n.getOrElse(0))
            (remaining, taken)
          }

        def view: F[List[A]] = state.get
      }
    }
}
```

or, to optimize the `offer` method to be O(1), we can use sth like [this](https://gist.github.com/ASRagab/82411aa5b78a9b06589d79238b15520e), so the `offer` would have O(1), and `take` would have amortized O(1) (O(n) if necessary but amortized over n `take` operations).

```scala
object ViewableQueue {

  case class QueueState[A](enqueueList: List[A], dequeueList: List[A])

  def emptyState[A]: QueueState[A] = QueueState(Nil, Nil)

  def make[F[_]: Concurrent, A]: F[ViewableQueue[F, A]] =
    Ref[F].of(emptyState[A]).map { state =>
      new ViewableQueue[F, A] {
        def offer(item: A): F[Unit] =
          state.update {
            case QueueState(enqueue, dequeue) => QueueState(item :: enqueue, dequeue)
          }

        private def ensureNonEmpty: F[Unit] =
          state.modify {
            case QueueState(enqueue, Nil) => (QueueState(Nil, enqueue.reverse), ())
            case s => (s, ())
          }

        def take: F[A] =
          for {
            _ <- ensureNonEmpty
            taken <- state.modify {
              case QueueState(enqueue, head :: tail) => (QueueState(enqueue, tail), head)
              case _ => throw new NoSuchElementException("take from an empty queue") // better handling needed, consider handling this later in effect
            }
          } yield taken

        def tryOfferN(items: List[A]): F[List[A]] =
          state.update {
            case QueueState(enqueue, dequeue) => QueueState(items.reverse ++ enqueue, dequeue)
          }.as(List.empty[A])

        def tryTakeN(n: Option[Int]): F[List[A]] =
          for {
            _ <- ensureDequeueListNonEmpty
            taken <- state.modify {
              case QueueState(enqueue, dequeue) =>
                val (taken, remaining) = dequeue.splitAt(n.getOrElse(0))
                (QueueState(enqueue, remaining), taken)
            }
          } yield taken

        def view: F[List[A]] =
          state.get.map {
            case QueueState(enqueue, dequeue) => dequeue ++ enqueue.reverse
          }
      }
    }
}
```

I think we don't need to use a Queue if we're going to keep all elements in the `List` anyway.

Reply:
Just to share, the test results for the second version you shared using `QueueState`.

**Test**
- Single Fiber
  - Insertion time: 519862 microseconds.
  - Retrieval time 50224 microseconds.
- Multiple fibers
  - Total insertion time for all fibers: 739888 microseconds
  - Retrieval time 26752 microseconds.

Reply:
Does it mean that the insertion is slightly slower but the retrieval is much faster? 🤔 Anyway, both ways would work fine IMO, so your call